src/Sarif.UnitTests/Visitors/InsertOptionalDataVisitorTests.cs

```diff
@@ -110,101 +110,230 @@ public void InsertOptionalDataVisitor_PersistsAll()
+
+            return run;
+        }
+
+        [Fact]
+        public void InsertOptionalDataVisitorTests_FlattensMessageStringsInResult()
+        {
+            Run run = CreateBasicRunForMessageStringLookupTesting();
+
+            run.Results.Add(
+                new Result
+                {
+                    RuleId = RuleId,
+                    RuleIndex = RuleIndex,
+                    Message = new Message
+                    {
+                        MessageId = UniqueGlobalMessageId
+                    }
+                });
+
+            run.Results.Add(
+                new Result
+                {
+                    RuleId = RuleId,
+                    RuleIndex = RuleIndex,
+                    Message = new Message
+                    {
+                        MessageId = UniqueRuleMessageId
+                    }
+                });
+
+
+            run.Results.Add(
+                new Result
+                {
+                    RuleId = RuleId,
+                    RuleIndex = RuleIndex,
+                    Message = new Message
+                    {
+                        MessageId = SharedMessageId
+                    }
+                });
+
+
+            var visitor = new InsertOptionalDataVisitor(OptionallyEmittedData.FlattenedMessages);
+            visitor.Visit(run);
+
+            run.Results[0].Message.Text.Should().Be(UniqueGlobalMessageValue);
+            run.Results[1].Message.Text.Should().Be(UniqueRuleMessageValue);
+
+            // Prefer rule-specific value in the event of a message id collision
+            run.Results[2].Message.Text.Should().Be(SharedKeyRuleMessageValue);
+        }
+
+        [Fact]
+        public void InsertOptionalDataVisitorTests_FlattensMessageStringsInNotification()
+        {
+            Run run = CreateBasicRunForMessageStringLookupTesting();
+
+            IList<Notification> toolNotifications = run.Invocations[0].ToolNotifications;
+            IList<Notification> configurationNotifications = run.Invocations[0].ConfigurationNotifications;
+
+            // Shared message id with no overriding rule id
+            toolNotifications.Add(
+                new Notification
+                {
+                    Id = NotificationId,
+                    Message = new Message {  MessageId = SharedMessageId}
+                });
+            configurationNotifications.Add(toolNotifications[0]);
+
+            // Shared message id with an overriding rule id. This message 
+            // should still be retrieved from the global strings table.
+            toolNotifications.Add(
+                new Notification
+                {
+                    Id = NotificationId,
+                    RuleIndex = RuleIndex,
+                    Message = new Message { MessageId = SharedMessageId }
+                });
+            configurationNotifications.Add(toolNotifications[1]);
+
+            toolNotifications.Add(
```

Comment:
>Add [](start = 30, length = 3)

Comment: "Notification that refers to a rule that does _not_ contain a message with the same id as the specified notification id. In this case it is no surprise that the message comes from the global string table."

... and then I think the test reads most naturally if you reverse the order of notifications 1 and 2.